src/items.rs

```diff
@@ -2072,19 +2099,30 @@ impl<'a> Rewrite for OpaqueType<'a> {
 
 impl Rewrite for ast::FnRetTy {
     fn rewrite(&self, context: &RewriteContext<'_>, shape: Shape) -> Option<String> {
+        self.rewrite_result(context, shape).ok()
+    }
+    fn rewrite_result(&self, context: &RewriteContext<'_>, shape: Shape) -> RewriteResult {
         match *self {
-            ast::FnRetTy::Default(_) => Some(String::new()),
+            ast::FnRetTy::Default(_) => Ok(String::new()),
             ast::FnRetTy::Ty(ref ty) => {
                 if context.config.version() == Version::One
                     || context.config.indent_style() == IndentStyle::Visual
                 {
-                    let inner_width = shape.width.checked_sub(3)?;
+                    let inner_width = shape
+                        .width
+                        .checked_sub(3)
+                        .max_width_error(shape.width, self.span())?;
                     return ty
-                        .rewrite(context, Shape::legacy(inner_width, shape.indent + 3))
+                        .rewrite_result(context, Shape::legacy(inner_width, shape.indent + 3))
                         .map(|r| format!("-> {}", r));
                 }
 
-                ty.rewrite(context, shape.offset_left(3)?)
+                // 3 = "-> "
+                let shape = shape
+                    .offset_left(3)
+                    .max_width_error(shape.width, self.span())?;
+
+                ty.rewrite_result(context, shape)
```

Comment:
nit: Since we're making changes here anyway, it might be nice to document all of the `+ 3` we've got. I'm pretty sure they all correspond to the length of `"-> "`. I know in a previous comment I suggested adding the comment `// 3 = "-> "`, but I think this would be an improvement on that.

```rust
                {
                    let arrow_width = "-> ".len();
                    let inner_width = shape
                        .width
                        .checked_sub(arrow_width)
                        .max_width_error(shape.width, self.span())?;
                    return ty
                        .rewrite_result(context, Shape::legacy(inner_width, shape.indent + arrow_width))
                        .map(|r| format!("-> {}", r));
                }

                let shape = shape
                    .offset_left(arrow_width)
                    .max_width_error(shape.width, self.span())?;

                ty.rewrite_result(context, shape)
```